File: app/src/main/java/net/tnose/app/trisquel/CameraFragment.java
package net.tnose.app.trisquel;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

/**
 * A fragment representing a list of Items.
 * <p/>
 * Activities containing this fragment MUST implement the {@link OnListFragmentInteractionListener}
 * interface.
 */
public class CameraFragment extends Fragment {

    // TODO: Customize parameter argument names
    private static final String ARG_COLUMN_COUNT = "column-count";
    // TODO: Customize parameters
    private int mColumnCount = 1;
    private OnListFragmentInteractionListener mListener;
    private ArrayList<CameraSpec> list;
    private MyCameraRecyclerViewAdapter cameraRecyclerViewAdapter;
    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public CameraFragment() {
    }

    // TODO: Customize parameter initialization
    @SuppressWarnings("unused")
    public static CameraFragment newInstance(int columnCount) {
        CameraFragment fragment = new CameraFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_COLUMN_COUNT, columnCount);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            mColumnCount = getArguments().getInt(ARG_COLUMN_COUNT);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_camera_list, container, false);
        TrisquelDao dao = new TrisquelDao(this.getContext());
        dao.connection();
        list = dao.getAllCameras();
        dao.close();

        // Set the adapter
        if (view instanceof RecyclerViewEmptySupport) {
            Context context = view.getContext();
            RecyclerViewEmptySupport recyclerView = (RecyclerViewEmptySupport) view;
            recyclerView.setEmptyMessage(getString(R.string.warning_cam_not_registered));
            recyclerView.setEmptyView(container.findViewById(R.id.empty_view));
            if (mColumnCount <= 1) {
                recyclerView.setLayoutManager(new LinearLayoutManager(context));
            } else {
                recyclerView.setLayoutManager(new GridLayoutManager(context, mColumnCount));
            }

            this.cameraRecyclerViewAdapter = new MyCameraRecyclerViewAdapter(list, mListener);
            recyclerView.setAdapter(cameraRecyclerViewAdapter);
        }
        return view;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            mListener = (OnListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void insertCamera(CameraSpec camera) {
        if (list != null) {
            int index = list.indexOf(camera);
            if (-1 == index) {
                list.add(0, camera);
                TrisquelDao dao = new TrisquelDao(this.getContext());
                dao.connection();
                long id = dao.addCamera(camera);
                dao.close();
                camera.id = (int)id;
                cameraRecyclerViewAdapter.notifyItemInserted(0);
            }
        }
    }

    public void updateCamera(CameraSpec camera) {
        if (list != null){
            for(int i = 0; i < list.size(); i++){
                if(list.get(i).id == camera.id){
                    list.remove(i);
                    list.add(i, camera);
                    TrisquelDao dao = new TrisquelDao(this.getContext());
                    dao.connection();
                    dao.updateCamera(camera);
                    dao.close();
                    cameraRecyclerViewAdapter.notifyItemChanged(i);
                }
            }
        }
    }

    public void deleteCamera(int id){
        if (list != null){
            for(int i = 0; i < list.size(); i++){
                CameraSpec c = list.get(i);
                if(c.id == id){
                    list.remove(i);
                    TrisquelDao dao = new TrisquelDao(this.getContext());
                    dao.connection();
                    if(c.type == 1){
                        dao.deleteLens(dao.getFixedLensIdByBody(id));
                    }
                    dao.deleteCamera(id);
                    dao.close();
                    cameraRecyclerViewAdapter.notifyItemRemoved(i);
                }
            }
        }
    }
    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     * <p/>
     * See the Android Training lesson <a href=
     * "http://developer.android.com/training/basics/fragments/communicating.html"
     * >Communicating with Other Fragments</a> for more information.
     */
    public interface OnListFragmentInteractionListener {
        // TODO: Update argument type and name
        void onListFragmentInteraction(CameraSpec item, boolean isLong);
    }
}

File: app/src/main/java/net/tnose/app/trisquel/CameraFragment.kt
package net.tnose.app.trisquel

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.unit.dp
import androidx.preference.PreferenceManager

class CameraFragment : androidx.fragment.app.Fragment() {
    private var mListener: OnListFragmentInteractionListener? = null
    private val cameraList = mutableStateListOf<CameraSpec>()
    private var scrollTargetIndex: Int? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        val dao = TrisquelDao(this.context)
        dao.connection()
        val list = dao.allCameras
        dao.close()

        cameraList.clear()
        cameraList.addAll(list)

        //ここでいいのか？
        val pref = PreferenceManager.getDefaultSharedPreferences(this.requireContext())
        val key  = pref.getInt("camera_sortkey", 0)
        changeSortKey(key)

        return ComposeView(requireContext()).apply {
            setContent {
                MaterialTheme {
                    CameraListScreen(
                        cameras = cameraList,
                        onItemClick = { mListener?.onListFragmentInteraction(it, false) },
                        onItemLongClick = { mListener?.onListFragmentInteraction(it, true) },
                        emptyMessage = getString(R.string.warning_cam_not_registered),
                        scrollTargetIndex = scrollTargetIndex,
                        onScrollConsumed = { scrollTargetIndex = null }
                    )
                }
            }
        }
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnListFragmentInteractionListener) {
            mListener = context
        } else {
            throw RuntimeException(context.toString() + " must implement OnListFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    fun changeSortKey(key: Int) {
        val sortedList = when (key) {
            0 -> cameraList.sortedByDescending { it.created }
            1 -> cameraList.sortedBy { it.manufacturer + " " + it.modelName }
            2 -> cameraList.sortedBy { it.mount }
            3 -> cameraList.sortedBy { it.format }
            else -> cameraList.toList()
        }
        cameraList.clear()
        cameraList.addAll(sortedList)
    }

    fun insertCamera(camera: CameraSpec) {
        val index = cameraList.indexOf(camera)
        if (-1 == index) {
            cameraList.add(0, camera)
            val dao = TrisquelDao(this.context)
            dao.connection()
            val id = dao.addCamera(camera)
            dao.close()
            camera.id = id.toInt()
            scrollTargetIndex = 0
        }
    }

    fun updateCamera(camera: CameraSpec) {
        val i = cameraList.indexOfFirst { it.id == camera.id }
        if (i != -1) {
            cameraList[i] = camera
            val dao = TrisquelDao(this.context)
            dao.connection()
            dao.updateCamera(camera)
            dao.close()
        }
    }

    fun deleteCamera(id: Int) {
        val i = cameraList.indexOfFirst { it.id == id }
        if (i != -1) {
            val c = cameraList[i]
            cameraList.removeAt(i)
            val dao = TrisquelDao(this.context)
            dao.connection()
            if (c.type == 1) {
                dao.deleteLens(dao.getFixedLensIdByBody(id))
            }
            dao.deleteCamera(id)
            dao.close()
        }
    }

    interface OnListFragmentInteractionListener {
        fun onListFragmentInteraction(item: CameraSpec, isLong: Boolean)
    }

    companion object {
        fun newInstance(columnCount: Int): CameraFragment {
            val fragment = CameraFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }
    }
}

@OptIn(ExperimentalFoundationApi::class)
@Composable
fun CameraListScreen(
    cameras: List<CameraSpec>,
    onItemClick: (CameraSpec) -> Unit,
    onItemLongClick: (CameraSpec) -> Unit,
    emptyMessage: String,
    scrollTargetIndex: Int?,
    onScrollConsumed: () -> Unit
) {
    if (cameras.isEmpty()) {
        Box(
            modifier = Modifier.fillMaxSize(),
            contentAlignment = Alignment.Center
        ) {
            Text(text = emptyMessage, style = MaterialTheme.typography.bodyLarge)
        }
    } else {
        val listState = rememberLazyListState()

        LaunchedEffect(scrollTargetIndex) {
            if (scrollTargetIndex != null && scrollTargetIndex < cameras.size) {
                listState.animateScrollToItem(scrollTargetIndex)
                onScrollConsumed()
            }
        }

        LazyColumn(
            state = listState,
            modifier = Modifier.fillMaxSize()
        ) {
            items(cameras, key = { it.id }) { camera ->
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .combinedClickable(
                            onClick = { onItemClick(camera) },
                            onLongClick = { onItemLongClick(camera) }
                        )
                        .padding(horizontal = 16.dp, vertical = 12.dp)
                ) {
                    Text(
                        text = "${camera.manufacturer} ${camera.modelName}",
                        style = MaterialTheme.typography.bodyLarge
                    )
                }
            }
        }
    }
}
